addons/enemybar/actionTracking.lua

```diff
@@ -0,0 +1,299 @@
+	end
+
+	-- find the action
+	local action_map = nil
+	if S{8,4}:contains(ai.category) then
+		action_map = res.spells[action_id]
+	elseif S{5,9}:contains(ai.category) then
+		action_map = res.items[action_id]
+	else
+		if actor.is_npc then
+			action_map = res.monster_abilities[action_id]
+		else
+			if ai.category == 6 then
+				action_map = res.job_abilities[action_id]
+			elseif S{3,7,11}:contains(ai.category) then
+				action_map = res.weapon_skills[action_id]
+			end
+		end
+	end
+	-- couldn't find the action, let's just give some debug output.
+	if not action_map then
+		action_map = {en='Unknown (id:'..action_id..')'}
+	end 
+
+	tracked_actions[ai.actor_id] = {actor=actor, target=target, ability=action_map, complete=S{4,11,3,5,6}:contains(ai.category), time=os.time()}
+end
+
+function track_debuffs(ai)
+	check_conflicting_debuffs(ai)
+	if S{8,7,6,9}:contains(ai.category) then return end
+
+	for i,t in ipairs(ai.targets) do
+		local target = windower.ffxi.get_mob_by_id(t.id)
+		if target and target.is_npc and not target.charmed then
+
+		    if S{2,252}:contains(t.actions[1].message) then
+		        local spell = ai.param
+		        local effect = res.spells[spell].status
+
+		        if effect then
+		            apply_debuff(target, effect, spell)
+		        end
+		        
+		    -- Non-damaging spells
+		    elseif S{75,236,237,268,270,271}:contains(t.actions[1].message) then
+		        local effect = t.actions[1].param
+		        local spell = ai.param
+
+		        apply_debuff(target, effect, spell)
+		    end
+		end
+	end
+end
+
+function apply_debuff(target, effect, spell)	
+    if not tracked_debuff[target.id] then
+        tracked_debuff[target.id] = {}
+    end
+
+    local overwrites = res.spells[spell].overwrites or {}
+    if not did_overwrite(target, spell, overwrites) then
+        return
+    end
+
+    tracked_debuff[target.id][effect] = {target=target,spell=spell,effect=effect,time=os.time(),duration=res.spells[spell].duration or 0,pos={x=target.x,y=target.y}}
+end
+
+function did_overwrite(target, new, t)
+    if not tracked_debuff[target.id] then
+        return true
+    end
+    
+    for effect, tracked in pairs(tracked_debuff[target.id]) do
+        local old = res.spells[tracked.spell].overwrites or {}
+        
+        -- Check if there isn't a higher priority debuff active
+        if table.length(old) > 0 then
+            for _,v in ipairs(old) do
+                if new == v then
+                    return false
+                end
+            end
+        end
+        
+        -- Check if a lower priority debuff is being overwritten
+        if table.length(t) > 0 then
+            for _,v in ipairs(t) do
+                if tracked.spell == v then
+                    tracked_debuff[target.id][effect] = nil
+                end
+            end
+        end
+    end
+    return true
+end
+
+function check_conflicting_debuffs(ai)
+	local actor = windower.ffxi.get_mob_by_id(ai.actor_id)
```

Comment:
Another function that calls `windower.ffxi.get_mob_by_id` which could get the entity via an argument instead.